File: algo-library/python/string/Z-algo.py
# Z[i] := S[0:|S|) と S[i:|S|) の最長共通接頭辞の長さ
def z_algo(S):
  Z = [0]*len(S)
  Z[0] = len(S)
  i,j = 1,0
  while i<len(S):
    # 一致する限り延長する
    while (i+j<len(S) and S[j]==S[i+j]): j += 1
    Z[i] = j
    
    # 一文字も一致しないならコピーできないので次に進む
    if j==0: 
      i += 1
      continue
    
    k = 1
    # k+Z[k]>j なら1回目の登場の際にはみ出してしまっている
    # k+Z[k]=j なら2回目の登場の際にはみ出す可能性がある
    while (k<j and k+Z[k]<j):
      Z[i+k] = Z[k]
      k += 1
    # まだ未確定なところまで i を進める
    i += k
    # j を i に合わせる
    j -= k
    
  return Z

File: algorithm_library/cpp-from-py/string/Z-algo.cpp
#include <bits/stdc++.h>
using namespace std;

template <class T>
vector<int> z_algo(const T& S) {
    int n = (int)S.size();
    vector<int> Z(n);
    if (n == 0) return Z;
    Z[0] = n;
    int i = 1, j = 0;
    while (i < n) {
        while (i + j < n && S[j] == S[i + j]) j++;
        Z[i] = j;
        if (j == 0) {
            i++;
            continue;
        }
        int k = 1;
        while (k < j && k + Z[k] < j) {
            Z[i + k] = Z[k];
            k++;
        }
        i += k;
        j -= k;
    }
    return Z;
}
